tarantool/connection.py

```diff
@@ -298,6 +329,46 @@ def connect_unix(self):
             self.connected = False
             raise NetworkError(e)
 
+    def wrap_socket_ssl(self):
+        '''
+        Wrap an existing socket with SSL socket.
+
+        :raise: NetworkError
+        :raise: `ssl.SSLError`
+        '''
+        if is_ssl_supported == False:
+            raise NetworkError("SSL is unsupported by the python.")
+
+        if hasattr(ssl, 'TLSVersion'):
+            # Since python 3.7
+            context = ssl.create_default_context(ssl.Purpose.SERVER_AUTH)
```

Comment:
Yes, I think you right. Fixed.

Reply:
I see that this code:

```python
# Since python 3.7
context = ssl.create_default_context(ssl.Purpose.SERVER_AUTH)
# Reset to default OpenSSL values
context.check_hostname = False
context.verify_mode = ssl.CERT_NONE
```

...was replaced with this one:

```python
# Since python 3.7
context = ssl.SSLContext(ssl.PROTOCOL_TLS)
```

The `create_default_context()` code from Python 3.10.2:

```python
def create_default_context(purpose=Purpose.SERVER_AUTH, *, cafile=None,
                           capath=None, cadata=None):
    """Create a SSLContext object with default settings.

    NOTE: The protocol and settings may change anytime without prior
          deprecation. The values represent a fair balance between maximum
          compatibility and security.
    """
    if not isinstance(purpose, _ASN1Object):
        raise TypeError(purpose)

    # SSLContext sets OP_NO_SSLv2, OP_NO_SSLv3, OP_NO_COMPRESSION,
    # OP_CIPHER_SERVER_PREFERENCE, OP_SINGLE_DH_USE and OP_SINGLE_ECDH_USE
    # by default.
    if purpose == Purpose.SERVER_AUTH:
        # verify certs and host name in client mode
        context = SSLContext(PROTOCOL_TLS_CLIENT)
        context.verify_mode = CERT_REQUIRED
        context.check_hostname = True 
    elif purpose == Purpose.CLIENT_AUTH:
        context = SSLContext(PROTOCOL_TLS_SERVER)
    else:
        raise ValueError(purpose)

    if cafile or capath or cadata:
        context.load_verify_locations(cafile, capath, cadata)
    elif context.verify_mode != CERT_NONE:
        # no explicit cafile, capath or cadata but the verify mode is
        # CERT_OPTIONAL or CERT_REQUIRED. Let's try to load default system
        # root CA certificates for the given purpose. This may fail silently.
        context.load_default_certs(purpose)
    # OpenSSL 1.1.1 keylog file
    if hasattr(context, 'keylog_filename'):
        keylogfile = os.environ.get('SSLKEYLOGFILE')
        if keylogfile and not sys.flags.ignore_environment:
            context.keylog_filename = keylogfile
    return context
```

I see the following differences in the behaviour:

1. No `context.load_default_certs()` call. It is the goal of the change.
2. `PROTOCOL_TLS` instead of `PROTOCOL_TLS_CLIENT`.

The [PROTOCOL_TLS_CLIENT](https://docs.python.org/3/library/ssl.html#ssl.PROTOCOL_TLS_CLIENT) differs this way:

> The protocol enables `CERT_REQUIRED` and `check_hostname` by default.

(And `PROTOCOL_TLS` is deprecated since Python 3.10.)

So all changes looks expected.

We can leave it as is or change a bit to avoid the deprecation warning (it is issued from `_ssl__SSLContext_impl()`):

```python
# Since python 3.7
context = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
# Reset to default OpenSSL values
context.check_hostname = False
context.verify_mode = ssl.CERT_NONE
```

To be honest, I'm in a doubt that accepting expired / another host certificate is the good default behaviour. OTOH, it looks incorrect to require providing a CA certificate from a client. What would be the ideal default in my opinion:

1. Check hostname in the server's certificate against provided `host`.
2. Check server's certificate expiration.
3. Don't validate server's certificate if `ssl_ca_file` is not passed.

However I'm not sure that we can achieve this behaviour using the `ssl` module. So, maybe, current behaviour is the best possible one.

Reply:
Hmm, I just now noticed that the implementation in tarantool does not check the host even if CA certificate is present:
https://github.com/tarantool/tarantool-ee/blob/master/src/lib/core/ssl.c#L141-L149

To test you can play with files from `test/data/`:

tnt1:
```
tarantool> box.cfg{ listen = { uri = 'localhost:3301', params = { transport = 'ssl', ssl_key_file = '../localhost.key', ssl_cert_file = '../invalidhost.crt', ssl_ca_file = '../ca.crt'} } }
```
tnt2:
```
tarantool> c = require('net.box').connect({uri = 'localhost:3301',params = {transport = 'ssl', ssl_key_file = '../localhost.key', ssl_cert_file = '../invalidhost.crt', ssl_ca_file = '../ca.crt'}})
tarantool> c
---
- peer_protocol_version: 3
  peer_uuid: 2d8c5692-0f67-4383-807d-603a763c851d
  host: localhost
  peer_protocol_features:
    transactions: true
    watchers: true
    error_extension: true
    streams: true
  schema_version: 78
  protocol: Binary
  state: active
  peer_version_id: 133632
  port: '3301'
...
```

Reply:
Thank you. The answer in Russian:
> Нет. Можно добавить проверку, но я не уверен, что она нужна. Возможно, стоит сделать ее опциональной (включать в зависимости от опций конфигурации).

```patch
@@ -376,14 +384,15 @@ class Connection(ConnectionInterface):
             if self.ssl_ca_file:
                 context.load_verify_locations(cafile=self.ssl_ca_file)
                 context.verify_mode = ssl.CERT_REQUIRED
-                if self.host:
-                    context.check_hostname = True
+                # A Tarantool implementation does not check hostname. We don't
+                # do that too. As a result we don't set here:
+                # context.check_hostname = True
```
I just delete the check and a comment because it's not obvious. We can always add the optional support of the feature later if needed.

Reply:
Okay.